flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/async/AbstractAsynchronousOperationHandlersTest.java

```diff
@@ -0,0 +1,350 @@
+import org.apache.flink.runtime.rest.messages.TriggerId;
+import org.apache.flink.runtime.rest.messages.TriggerIdPathParameter;
+import org.apache.flink.runtime.rest.messages.queue.QueueStatus;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.TestingRestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.runtime.webmonitor.retriever.TestingGatewayRetriever;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.annotation.Nullable;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for the {@link AbstractAsynchronousOperationHandlers}.
+ */
+public class AbstractAsynchronousOperationHandlersTest extends TestLogger {
+
+	private static final CompletableFuture<String> localRestAddress = CompletableFuture.completedFuture("localhost");
+
+	private static final Time timeout = Time.seconds(10L);
+
+	private TestingAsynchronousOperationHandlers testingAsynchronousOperationHandlers;
+
+	private TestingAsynchronousOperationHandlers.TestingTriggerHandler testingTriggerHandler;
+
+	private TestingAsynchronousOperationHandlers.TestingStatusHandler testingStatusHandler;
+
+	private TestingGatewayRetriever<RestfulGateway> testingGatewayRetriever;
+
+	@Before
+	public void setup() {
+		testingGatewayRetriever = new TestingGatewayRetriever<>();
+		testingAsynchronousOperationHandlers = new TestingAsynchronousOperationHandlers();
+
+		testingTriggerHandler = testingAsynchronousOperationHandlers.new TestingTriggerHandler(
+			localRestAddress,
+			testingGatewayRetriever,
+			timeout,
+			Collections.emptyMap(),
+			TestingTriggerMessageHeaders.INSTANCE);
+
+		testingStatusHandler = testingAsynchronousOperationHandlers.new TestingStatusHandler(
+			localRestAddress,
+			testingGatewayRetriever,
+			timeout,
+			Collections.emptyMap(),
+			TestingStatusMessageHeaders.INSTANCE);
+	}
+
+	/**
+	 * Tests the triggering and successful completion of an asynchronous operation.
+	 */
+	@Test
+	public void testOperationCompletion() throws Exception {
+		final CompletableFuture<String> savepointFuture = new CompletableFuture<>();
+		final TestingRestfulGateway testingRestfulGateway = new TestingRestfulGateway.Builder()
+			.setTriggerSavepointFunction((JobID jobId, String directory) -> savepointFuture)
+			.build();
+
+		testingGatewayRetriever.setGateway(testingRestfulGateway);
```

Comment:
Sometimes you set the gateway in the test retriever, sometimes you don't (e.g., `testUnknownTriggerId`). We don't rely on `testingGatewayRetriever` in the tests because `handleRequest` expects the `RestfulGateway` as an argument.

Reply:
You're right. Will remove it.